android-components/components/feature/top-sites/src/main/java/mozilla/components/feature/top/sites/DefaultTopSitesStorage.kt

```diff
@@ -105,14 +107,17 @@ class DefaultTopSitesStorage(
                     .filter { providerConfig.providerFilter?.invoke(it) ?: true }
                     .take(numSitesRequired)
                     .take(providerConfig.maxThreshold - pinnedSites.size)
+                pinnedExclusions = topSitesProvider.getPinnedExclusions()
                 topSites.addAll(providerTopSites)
                 numSitesRequired -= providerTopSites.size
             } catch (e: Exception) {
                 logger.error("Failed to fetch top sites from provider", e)
             }
         }
 
-        topSites.addAll(pinnedSites)
+        topSites.addAll(
+            pinnedSites.filter { !pinnedExclusions.contains(it.url.getRepresentativeSnippet()) },
+        )
```

Comment:
Consider the thoughts in https://github.com/mozilla-mobile/firefox-android/pull/898/files#r1203355929 first, but one of the bugs I see here is that we are potentially no longer displaying the right number of top sites because we are filtering after subtracting the `numSitesRequired` above [L97](https://github.com/mozilla-mobile/firefox-android/pull/898/files#diff-81df006dcf09133e7186c24352b59822b7b76bf49efcaf4d36ded81cc181924eR97). 

For example, if we specified that we want to display 8 top sites, we might end up only displaying 7 if one of them got filtered out and we did not adjust the count of `numSitesRequired`.

Reply:
Is it acceptable that I adjust `numSitesRequired` before we get frecent top sites, so that frecent top sites can fill up?